# Project Flow:
1. Problem statement
2. Data Gathering
3. Data Preprocessing:
   i. tokenization
   ii. lowercase conversion
   iii. removal of stopwords
   iv. appyling lemmatization/stemming
   
4. Vectorization:
   i. CountVectorizer
   ii. TF-IDF
   
5. Model Building:
   i. Model Object initialization
   ii. train and test model
   
6. Model Evaluation:
   i. Accuracy score
   ii. Confusion Matrix
   iii. Classification report
   
7. Model Deployment
8. Prediction on data


# 1. Problem statement
# Fake news classification by using Natural Language Processing.
Fake news detection is a hot topic nowsdays in the field of natural language processing. We get the new through several mediums throughout the day in our daily life, but sometimes some news are fake. So, inorder to classify whether the news is fake or real, we create a model which will predicts whether the given news id real or fake.

# Required Libraries

In [1]:
import pandas as pd
import numpy as np
import re
import nltk

from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# 2. Data Gathering

In [2]:
data = pd.read_csv('D:\\Projects\\NLP_Projects\\News_dataset.csv')
data.head()

id                                              title              author  \
0   0  House Dem Aide: We Didn’t Even See Comey’s Let...       Darrell Lucus   
1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...     Daniel J. Flynn   
2   2                  Why the Truth Might Get You Fired  Consortiumnews.com   
3   3  15 Civilians Killed In Single US Airstrike Hav...     Jessica Purkiss   
4   4  Iranian woman jailed for fictional unpublished...      Howard Portnoy   

                                                text  label  
0  House Dem Aide: We Didn’t Even See Comey’s Let...      1  
1  Ever get the feeling your life circles the rou...      0  
2  Why the Truth Might Get You Fired October 29, ...      1  
3  Videos 15 Civilians Killed In Single US Airstr...      1  
4  Print \nAn Iranian woman has been sentenced to...      1

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20800 entries, 0 to 20799
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      20800 non-null  int64 
 1   title   20242 non-null  object
 2   author  18843 non-null  object
 3   text    20761 non-null  object
 4   label   20800 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 812.6+ KB


There are total 20800 rows of data and 5 columns.
And, there are 3 colums which are object datatype.

In [4]:
data.isnull().sum()

id           0
title      558
author    1957
text        39
label        0
dtype: int64

isnull() is used to calculate the number of null rows.
According to above, there are total of 558 null values in title column and 1957 null values in author column and very less null values present in text column

# droping the null values

In [5]:
data = data.dropna()

In [6]:
data.isnull().sum()

id        0
title     0
author    0
text      0
label     0
dtype: int64

# Reseting index flows

In [7]:
data.reset_index(inplace=True)
data.head()

index  id                                              title  \
0      0   0  House Dem Aide: We Didn’t Even See Comey’s Let...   
1      1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...   
2      2   2                  Why the Truth Might Get You Fired   
3      3   3  15 Civilians Killed In Single US Airstrike Hav...   
4      4   4  Iranian woman jailed for fictional unpublished...   

               author                                               text  \
0       Darrell Lucus  House Dem Aide: We Didn’t Even See Comey’s Let...   
1     Daniel J. Flynn  Ever get the feeling your life circles the rou...   
2  Consortiumnews.com  Why the Truth Might Get You Fired October 29, ...   
3     Jessica Purkiss  Videos 15 Civilians Killed In Single US Airstr...   
4      Howard Portnoy  Print \nAn Iranian woman has been sentenced to...   

   label  
0      1  
1      0  
2      1  
3      1  
4      1

# Droping unnecessary column

In [8]:
data = data.drop(['id','author','text'],axis=1)
data.head()

index                                              title  label
0      0  House Dem Aide: We Didn’t Even See Comey’s Let...      1
1      1  FLYNN: Hillary Clinton, Big Woman on Campus - ...      0
2      2                  Why the Truth Might Get You Fired      1
3      3  15 Civilians Killed In Single US Airstrike Hav...      1
4      4  Iranian woman jailed for fictional unpublished...      1

# 3. Data Preprocessing

# 3.i Tokenization

In [9]:
sample_data = 'The quick brown fox jumps over the lazy dog'
sample_data = sample_data.split()
sample_data

['The', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']

# 3.ii Lowercase conversion

In [10]:
sample_data = [data.lower() for data in sample_data]
sample_data

['the', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']

# 3.iii Applying stopwords

In [11]:
stopwords = stopwords.words('english')
print(stopwords[0:10])
print(len(stopwords))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]
179


In [12]:
sample_data = [data for data in sample_data if data not in stopwords]
print(sample_data)
len(sample_data)

['quick', 'brown', 'fox', 'jumps', 'lazy', 'dog']


6

# 3.iv applying stemming and Lemmatization

In [13]:
ps = PorterStemmer()
sample_data_stemming = [ps.stem(data) for data in sample_data]
print(sample_data_stemming)

['quick', 'brown', 'fox', 'jump', 'lazi', 'dog']


In [14]:
lm = WordNetLemmatizer()
sample_data_lemma = [lm.lemmatize(data) for data in sample_data]
print(sample_data_lemma)

['quick', 'brown', 'fox', 'jump', 'lazy', 'dog']


In [15]:
lm = WordNetLemmatizer()
corpus = []
for i in range (len(data)):
    review = re.sub('^a-zA-Z0-9',' ',data['title'][i])
    review = review.lower()
    review = review.split()
    review = [lm.lemmatize(x) for x in review if x not in stopwords]
    review = " ".join(review)
    corpus.append(review)

In [16]:
len(corpus)

18285

In [17]:
data['title'][0]

'House Dem Aide: We Didn’t Even See Comey’s Letter Until Jason Chaffetz Tweeted It'

In [18]:
corpus[0]

'house dem aide: didn’t even see comey’s letter jason chaffetz tweeted'

# 4. Vectorization

In [19]:
tf = TfidfVectorizer()
x = tf.fit_transform(corpus).toarray()
x

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [20]:
y = data['label']
y.head()

0    1
1    0
2    1
3    1
4    1
Name: label, dtype: int64

# spliting data in train and test

In [21]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=12, stratify=y)

In [22]:
len(x_train),len(y_train)

(12799, 12799)

In [23]:
len(x_test),len(y_test)

(5486, 5486)

# Model building

In [24]:
rf = RandomForestClassifier()
rf.fit(x_train,y_train)

RandomForestClassifier()

# Model Evaluation

In [25]:
y_pred = rf.predict(x_test)
accuracy = accuracy_score(y_test,y_pred)
accuracy

0.9327378782355086

In [33]:
class Evaluation:
    def __init__(self,model,x_train,x_test,y_train,y_test):
        self.model = model
        self.x_train = x_train
        self.x_test = x_test
        self.y_train = y_train
        self.y_test = y_test
        
    def train_evaluation(self):
        y_pred_train = self.model.predict(self.x_train)
        
        acc_train = accuracy_score(self.y_train,y_pred_train)
        print("Accuracy score on training dataset",acc_train)
        
        confusion_train = confusion_matrix(self.y_train,y_pred_train)
        print("Confusion matrix on training dataset :\n",confusion_train)
        
        classification_repo = classification_report(self.y_train,y_pred_train)
        print("Classification report on Training dataset :\n",classification_repo)
        
    def test_evaluation(self):
        y_pred_test = self.model.predict(self.x_test)
        
        acc_test = accuracy_score(self.y_test,y_pred_test)
        print("Accuracy score on training dataset",acc_test)
        
        confusion_test = confusion_matrix(self.y_test,y_pred_test)
        print("Confusion matrix on training dataset :\n",confusion_test)
        
        classification_repo = classification_report(self.y_test,y_pred_test)
        print("Classification report on Training dataset :\n",classification_repo)
        

In [34]:
Evaluation(rf,x_train,x_test,y_train,y_test).train_evaluation()

Accuracy score on training dataset 1.0
Confusion matrix on training dataset :
 [[7252    0]
 [   0 5547]]
Classification report on Training dataset :
               precision    recall  f1-score   support

           0       1.00      1.00      1.00      7252
           1       1.00      1.00      1.00      5547

    accuracy                           1.00     12799
   macro avg       1.00      1.00      1.00     12799
weighted avg       1.00      1.00      1.00     12799



In [35]:
Evaluation(rf,x_train,x_test,y_train,y_test).test_evaluation()

Accuracy score on training dataset 0.9327378782355086
Confusion matrix on training dataset :
 [[2801  308]
 [  61 2316]]
Classification report on Training dataset :
               precision    recall  f1-score   support

           0       0.98      0.90      0.94      3109
           1       0.88      0.97      0.93      2377

    accuracy                           0.93      5486
   macro avg       0.93      0.94      0.93      5486
weighted avg       0.94      0.93      0.93      5486



# Prediction

In [36]:
class Preprocessing:
    def __init__(self,data):
        self.data= data
    
    def text_preprocessing_user(self):
        lm = WordNetLemmatizer()
        pred_data = [self.data]
        preprocess_data = []
        for data in pred_data:
            review = re.sub('^a-zA-Z0-9',' ',data)
            review = review.lower()
            review = review.split()
            review = [lm.lemmatize(x) for x in review if x not in stopwords]
            review = " ".join(review)
            preprocess_data.append(review)
        return preprocess_data

In [37]:
data['title'][1]

'FLYNN: Hillary Clinton, Big Woman on Campus - Breitbart'

In [39]:
data = 'FLYNN: Hillary Clinton, Big Woman on Campus - Breitbart'
Preprocessing(data).text_preprocessing_user()

['flynn: hillary clinton, big woman campus - breitbart']

In [47]:
class Prediction:
    def __init__(self,pred_data,model):
        self.pred_data = pred_data
        self.model = model
        
    def prediction_model(self):
        preprocess_data = Preprocessing(self.pred_data).text_preprocessing_user()
        data = tf.transform(preprocess_data)
        prediction = self.model.predict(data)
        
        if prediction [0]==0:
            return "The News is fake"
        else:
            return "The News is Real"
        

In [48]:
data='FLYNN: Hillary Clinton, Big Woman on Campus - Breitbart'
Prediction(data,rf).prediction_model()

'The News is fake'

In [49]:
user_data = 'Iranian woman jailed for fictional unpublished.'
Prediction(user_data,rf).prediction_model()

'The News is Real'